bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,494 @@
+ */
+
+#include "bgpd/bgp_conditional_adv.h"
+
+const char *get_afi_safi_str(afi_t afi, safi_t safi, bool for_json);
+
+/* We just need bgp_dest node matches with filter prefix. So no need to
+ * traverse each path here. */
+struct bgp_dest *bgp_dest_matches_filter_prefix(struct bgp_table *table,
+						struct filter *filter)
+{
+	uint32_t check_addr;
+	uint32_t check_mask;
+	struct in_addr mask;
+	struct bgp_dest *dest = NULL;
+	struct bgp_path_info *pi = NULL;
+	const struct prefix *dest_p = NULL;
+	struct filter_cisco *cfilter = NULL;
+	struct filter_zebra *zfilter = NULL;
+
+	if (!table || !filter)
+		return dest;
+
+	if (filter->cisco) {
+		cfilter = &filter->u.cfilter;
+		for (dest = bgp_table_top(table); dest;
+		     dest = bgp_route_next(dest)) {
+			dest_p = (struct prefix *)bgp_dest_get_prefix(dest);
+			if (!dest_p)
+				continue;
+			pi = bgp_dest_get_bgp_path_info(dest);
+			if (!pi)
+				continue;
+			check_addr = dest_p->u.prefix4.s_addr
+				     & ~cfilter->addr_mask.s_addr;
+			if (memcmp(&check_addr, &cfilter->addr.s_addr, 4) != 0)
+				continue;
+			if (cfilter->extended) {
+				masklen2ip(dest_p->prefixlen, &mask);
+				check_mask = mask.s_addr
+					     & ~cfilter->mask_mask.s_addr;
+				if (memcmp(&check_mask, &cfilter->mask.s_addr,
+					   4)
+				    != 0)
+					continue;
+			}
+			return dest;
+		}
+	} else {
+		zfilter = &filter->u.zfilter;
+		for (dest = bgp_table_top(table); dest;
+		     dest = bgp_route_next(dest)) {
+			dest_p = bgp_dest_get_prefix(dest);
+			if (!dest_p)
+				continue;
+			pi = bgp_dest_get_bgp_path_info(dest);
+			if (!pi)
+				continue;
+			if ((zfilter->prefix.family != dest_p->family)
+			    || (zfilter->exact
+				&& (zfilter->prefix.prefixlen
+				    != dest_p->prefixlen)))
+				continue;
+			else if (!prefix_match(&zfilter->prefix, dest_p))
+				continue;
+			else
+				return dest;
+		}
+	}
+	return NULL;
+}
+
+enum route_map_cmd_result_t
+bgp_check_rmap_prefixes_in_bgp_table(struct bgp_table *table,
+				     struct route_map *rmap)
+{
+	afi_t afi;
+	struct access_list *alist = NULL;
+	struct filter *alist_filter = NULL;
+	struct bgp_dest *dest = NULL;
+	struct route_map_rule *match = NULL;
+	enum route_map_cmd_result_t ret = RMAP_NOOP;
+
+	if (!table || !is_rmap_valid(rmap))
+		return ret;
+
+	/* If several match commands are configured, all must succeed for a
+	 * given route in order for that route to match the clause (logical AND)
+	 */
+	for (match = rmap->head->match_list.head; match; match = match->next) {
+
+		if (!match->cmd || !match->cmd->str || !match->value)
+			continue;
+
+		ret = RMAP_NOMATCH;
+
+		afi = get_afi_from_str(match->cmd->str);
+		if (afi == AFI_MAX)
+			return ret;
+
+		alist = access_list_lookup(afi, (char *)match->value);
+		if (!alist)
+			return ret;
+
+		/* If a match command refers to several objects in one
+		 * command either of them should match (i.e logical OR) */
+		FOREACH_ACCESS_LIST_FILTER(alist, alist_filter) {
```

Comment:
`route_map_apply()` wouldn't be a good candidate here?

Reply:
Sorry, route-map-apply() cannot be used as it is. because here the lookup is reverse.
Route-map : 'For each prefix in BGP table', we check for the best match in route-map & 
                      if there is no best match, route-map action is to deny the prefix and withdraw the route. 
advertise-map : 'For each prefix in route-map match filter' if there is a match or no match in BGP table, 
                          action will be taken based on prefix match, condition(exist/non-exist) and previous state.
                          and if there is a config change or not. this is to avoid sending duplicate advertisements.
 Also every 60 seconds we need to scan for table entries. 
So to avoid additional look ups and unnecessary checks written new functions which are necessary to be checked.